sync/peerset/peer_set.go

```diff
@@ -59,18 +60,31 @@ func (ps *PeerSet) FindSession(id int) *Session {
 	return nil
 }
 
-func (ps *PeerSet) HasAnyValidSession() bool {
+func (ps *PeerSet) NumberOfOpenSessions() int {
+	ps.lk.RLock()
+	defer ps.lk.RUnlock()
+
+	ps.removeExpiredSessions()
+
+	return len(ps.sessions)
+}
+
+func (ps *PeerSet) HasAnyOpenSession() bool {
 	ps.lk.Lock()
 	defer ps.lk.Unlock()
 
+	ps.removeExpiredSessions()
+
+	return len(ps.sessions) != 0
+}
+
+func (ps *PeerSet) removeExpiredSessions() {
```

Comment:
since both `HasAnyOpenSession` and `NumberOfOpenSessions`  are tempering with `ps.sessions` i don't see why there lucks are different i think both need Read and Write protective lock

Reply:
Thanks, Fixed.